### ISSUE TITLE
aqojHow can we improve the search results?hukonteuqlwjkmqlwjkö<

### ISSUE THREAD

**Issue by samhedin** [samhedin/rustdoc-to-org#2]

**Comment by samhedin** [comment]
@snyball you seem familiar with many of these tools. Do you have any ideas for what we could consider? Just off the top of my head there is fzf, but I have no experience with it.

**Comment by snyball** [comment]
@samhedin I think the only way you're going to get any good results is to parse the `.org` files and create an index, then probably make a rule that says:

- If there are upper-case characters, assume the user wants to look at the documentation for a struct or a trait.
- Otherwise, only search for functions/macros.

The index should probably just be a long text file with lines like these:

function_names.txt
```
name_of_function
name_of_other_function
maybe_also_macros!
```

With an accompanying file for paths:

function_paths.txt
```
path/to/file/containing/name_of_function:<line-number>
path/to/file/containing/name_of_other_function:<line-number>
path/to/file/containing/mayebe_also_macros!:<line-number>
```

Then you can get pretty fast searching by searching for a function in `function_names.txt`, extracting the line number of the hit, then looking up the line number in `function_paths.txt`, and opening the `.org` file on the specified `<line-number>`

The reason we split this up like this is so that we can use ripgrep to do the actual search, a completely elisp-based search might be quite a bit slower.

Note: It might also be a good idea to load the `function_paths.txt` file into an elisp vector for fast lookup, but you'll have to see how long that takes to load, and how much memory it ends up consuming (probably not a lot, but I've never worked with huge elisp vectors.)

Then you'll do something similar for structs and traits.

**Comment by snyball** [comment]
@samhedin Then, we probably also want it integrated nicely with LSP, by modifying `lsp-describe-thing-at-point` to work with rustdoc:

```elisp
(defun lsp-describe-thing-at-point ()
  "Display the type signature and documentation of the thing at
point."
  (interactive)
  (let ((contents (-some->> (lsp--text-document-position-params)
                    (lsp--make-request "textDocument/hover")
                    (lsp--send-request)
                    (lsp:hover-contents))))
    (if (and contents (not (equal contents "")))
        (let ((lsp-help-buf-name "*lsp-help*"))
          (with-current-buffer (get-buffer-create lsp-help-buf-name)
            (with-help-window lsp-help-buf-name
              (insert (string-trim-right (lsp--render-on-hover-content contents t))))))
      (lsp--info "No content at point."))))
```

**Comment by samhedin** [comment]
> I think the only way you're going to get any good results is to parse the `.org` files and create an index, then probably make a rule that says:

Yeah, I was "worried" that this would turn out to be the solution since it's also the one that I expect would require the most work. I considered a solution with SQLite a while back but felt like the regex solution was fine for the moment. I'm not sure how it matches up with a text-based solution, probably need to investigate both. 

>Also, is it possible to generate something like this with the pandoc filter? It should be able to spit out information to files, that can then be joined afterwards (otherwise you'll have a whole bunch of race-conditions.)

I doubt it's possible to get information such as line number from pandoc. We can do side effects and extract information, but it would be things like "this file contains this function/struct/...". It might be easier to build the index afterwards by going through the processed files instead of trying to do everything at once in Lua.

**Comment by snyball** [comment]
@samhedin It looks like you can get the org-path though, so if instead of line numbers, we store something like "Heading>Subheading>Subsubheading" then org-mode knows how to navigate there.

**Comment by samhedin** [comment]
> @samhedin Then, we probably also want it integrated nicely with LSP, by modifying `lsp-describe-thing-at-point` to work with rustdoc:
> 
> ```emacs-lisp
> (defun lsp-describe-thing-at-point ()
>   "Display the type signature and documentation of the thing at
> point."
>   (interactive)
>   (let ((contents (-some->> (lsp--text-document-position-params)
>                     (lsp--make-request "textDocument/hover")
>                     (lsp--send-request)
>                     (lsp:hover-contents))))
>     (if (and contents (not (equal contents "")))
>         (let ((lsp-help-buf-name "*lsp-help*"))
>           (with-current-buffer (get-buffer-create lsp-help-buf-name)
>             (with-help-window lsp-help-buf-name
>               (insert (string-trim-right (lsp--render-on-hover-content contents t))))))
>       (lsp--info "No content at point."))))
> ```

I looked into this too and I'm no longer sure of how much use we can get out of LSP - at least not without digging around in rust-analyzer. Taking a look at `lsp-rust-analyzer-syntax-tree` does not reveal much more information about something than what is already visible inside the document.

**Comment by samhedin** [comment]
> @samhedin It looks like you can get the org-path though, so if instead of line numbers, we store something like "Heading>Subheading>Subsubheading" then org-mode knows how to navigate there.

Neat, this information is available. Still not sure if it's easier than just looking through it later, but that might be because of my general skepticism towards Lua.

**Comment by snyball** [comment]
The information is pretty easy to extract with lsp, with the cursor on `something.clone()`

```
ELISP> (setq thing
             (-some->> (lsp--text-document-position-params)
                    (lsp--make-request "textDocument/hover")
                    (lsp--send-request)
                    (lsp:hover-contents)))
#<hash-table equal 2/2 0x158b1b97e393>
ELISP> (print thing)
#s(hash-table size 2 test equal rehash-size 1.5 rehash-threshold 0.8125 data ("kind" "markdown" "value" "rust
core::clone::Clone
fn clone(&self) -> Self

Returns a copy of the value.

# Examples

let hello = \"Hello\"; // &str implements Clone

assert_eq!(\"Hello\", hello.clone());"))

#<hash-table equal 2/2 0x158b1b97e393>
```

Also, SQLite would work just fine, I'm just biased towards plain-text.

**Comment by samhedin** [comment]
Sweet! I'll start thinking about the index and see where it leads.

**Comment by samhedin** [comment]
Rustdoc now uses `lsp-thing-at-point` where possible.
Hover over `Some` -> search for `std::option::Option::Some`. In order to use this, the lua filter inserts the complete crate name in the org file, so a header earlier named `Some` is now called `std::option::Option::Some`.
Having the crate name repeated constantly in the org doc is clearly redundant, and in the end, this is what the index will be for. But this way I could try how this works and take a small step at a time.
